File: src/main/scala/venix/hookla/services/DiscordMessageService.scala
package venix.hookla.services

import ackcord.data.OutgoingEmbed
import com.twitter.finagle.{Http, Service, http}
import com.twitter.finagle.http.Response
import com.twitter.io.Buf
import com.twitter.util.Future
import io.circe.syntax._
import venix.hookla.models.DiscordWebhook
import venix.hookla.types.OutgoingWebhookPayload

class DiscordMessageService {
  private val discordClient: Service[http.Request, http.Response] =
    Http.client.withSessionQualifier.noFailFast.withSessionQualifier.noFailureAccrual
      .withTls("discord.com")
      .newService("discord.com:443")

  def sendMessageToDiscord(discordWebhook: DiscordWebhook, embed: OutgoingEmbed): Future[Response] = {
    val payload = OutgoingWebhookPayload(List(embed))
    val request =
      http.Request(http.Method.Post, s"/api/webhooks/${discordWebhook.discordWebhookId}/${discordWebhook.discordWebhookToken}")

    println("Sending following JSON to Discord...")
    println(payload.asJson)

    request.write(payload.asJson.toString())
    request.headerMap.add("Content-Type", "application/json")

    discordClient(request) onSuccess { res: Response =>
      res.reader.read() map { m =>
        val Buf.Utf8(str) = m.get
        println(str)
      }

      println("success", res)
    } onFailure { ex: Throwable => println(ex) }
  }

}

File: cpp/src/services/DiscordMessageService.cpp
#include "DiscordMessageService.h"
#include <httplib.h>
#include <nlohmann/json.hpp>
#include <spdlog/spdlog.h>

DiscordMessageService::DiscordMessageService() = default;

void DiscordMessageService::sendMessageToDiscord(const DiscordWebhook& hook, const DiscordEmbed& embed) {
    nlohmann::json payload = {{"embeds", nlohmann::json::array({embedToJson(embed)})}};
    const std::string body = payload.dump();
    const std::string path = "/api/webhooks/" + hook.discordWebhookId + "/" + hook.discordWebhookToken;

    spdlog::debug("Sending to Discord: {}", body);

    httplib::SSLClient client("discord.com");
    client.set_connection_timeout(5);
    client.set_read_timeout(10);

    auto res = client.Post(path, body, "application/json");

    if (!res)
        spdlog::error("Discord request failed: no response");
    else if (res->status < 200 || res->status >= 300)
        spdlog::error("Discord returned HTTP {}: {}", res->status, res->body);
    else
        spdlog::debug("Discord webhook sent successfully (HTTP {})", res->status);
}
